File: aux/tests/matrix/boolean_masks.py
# column masks
result = []
res = 1 | (1 << 8) | (1 << 16) | (1 << 24) | (1 << 32) | (1 << 40) | (1 << 48) | (1 << 56)
for j in range (8):
    result.append (str(res << j)+"LLU")
print ("column masks")
print (", ".join(result))

# row masks
result = []
res = 1 | (1 << 1) | (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5) | (1 << 6) | (1 << 7)
for j in range (8):
    result.append (str(res << (j*8))+"LLU")
print ("\nrow masks")
print (", ".join(result))

# traspose masks 
l = [56, 48, 40, 32, 24, 16, 8, 0, 1, 2, 3, 4, 5, 6, 7]
a = zip (l, (8 - abs(x) for x in range (-7, 8)))
result = []
for i in a:
    r = i[0]
    res = 0
    for j in range (0, 9 * i[1], 9):
        res |= 1 << (r + j)
    result.append (str(res) + "LLU")
print ("\ntranspose masks")
print (", ".join(result))

File: aux/tests/matrix/boolean_masks.cpp
#include <iostream>
#include <ranges>
#include <vector>
#include <cstdint>
#include <string>

void printm (uint64_t a) {
    uint64_t one = 1;
    std::cout << std::endl;
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            std::cout << !!(a & (one << ((i << 3) + j))) << " ";
        }
        std::cout << std::endl;
    }
}

int main () {
    std::cout << "traspose masks" << std::endl;
    std::vector<int> vec = {
        56, 48, 40, 32, 24, 16, 8, 1, 2, 3, 4, 5, 6, 7
    };
    std::vector<int> ind = {
        1, 2, 3, 4, 5, 6, 7, 7, 6, 5, 4, 3, 2, 1
    };
    uint64_t res, one = 1;
    for (const auto & [a, b]: std::views::zip (vec, ind)) {
        res = 0;
        for (int i = 0; i < b; i++) {
            res |= one << (a + i*9);
        }
        std::cout << std::to_string(res) << "ULL, "; 
    }
    std::cout << std::endl;
    
    std::cout << std::endl << "row masks" << std::endl;
    std::vector<uint64_t> vec2;
    vec2.clear ();
    res = 255;
    for (int i = 0; i < 8; i++) {
        vec2.push_back (res);
        std::cout << std::to_string(res) << "ULL, ";
        res <<= 8;
    }
    
    std::cout << std::endl << "column masks" << std::endl;
    res = 0;
    res = one | (one << 8) | (one << 16) | (one << 24) | (one << 32) | (one << 40) | (one << 48) | (one << 56);
    for (int i = 0; i < 8; i++) {
        std::cout << res << "ULL, ";
        res <<= 1;
    }

    std::vector<uint64_t> vec1;
    vec1.clear ();
    std::cout << std::endl << "logarithm masks" << std::endl;
    vec1.push_back (vec2[7] | vec2[6] | vec2[5] | vec2[4]);
    vec1.push_back (vec2[3] | vec2[2]);
    vec1.push_back (vec2[1]);
    std::cout << vec1[0] << "ULL, ";
    std::cout << vec1[1] << "ULL, ";
    std::cout << vec1[2] << "ULL, ";

    printm(vec1[0]);
    printm(vec1[1]);
    printm(vec1[2]);

    std::cout << std::endl;

    return 0;
}
